src/metabase/api/collection.clj

```diff
@@ -428,23 +431,40 @@
 
 (api/defendpoint PUT "/:id"
   "Modify an existing Collection, including archiving or unarchiving it, or moving it."
-  [id, :as {{:keys [name color description archived parent_id], :as collection-updates} :body}]
-  {name        (s/maybe su/NonBlankString)
-   color       (s/maybe collection/hex-color-regex)
-   description (s/maybe su/NonBlankString)
-   archived    (s/maybe s/Bool)
-   parent_id   (s/maybe su/IntGreaterThanZero)}
+  [id, :as {{:keys [name color description archived parent_id type type_sub_tree], :as collection-updates} :body}]
+  {name          (s/maybe su/NonBlankString)
+   color         (s/maybe collection/hex-color-regex)
+   description   (s/maybe su/NonBlankString)
+   archived      (s/maybe s/Bool)
+   parent_id     (s/maybe su/IntGreaterThanZero)
+   type          collection/Type
+   type_sub_tree (s/maybe s/Bool)}
   ;; do we have perms to edit this Collection?
   (let [collection-before-update (api/write-check Collection id)]
     ;; if we're trying to *archive* the Collection, make sure we're allowed to do that
     (check-allowed-to-archive-or-unarchive collection-before-update collection-updates)
+    (when (or (and (contains? collection-updates :type)
+                   (not= type (:type collection-before-update)))
+              type_sub_tree)
+      (api/check-403 (and api/*is-superuser?*
+                          ;; pre-update of model checks if the collection is a personal collection and rejects changes
+                          ;; to type, but it doesn't check if it is a sub-collection of a personal one so we add that
+                          ;; here
+                          (not (collection/is-personal-collection-or-descendant-of-one? collection-before-update)))))
```

Comment:
I've actually added a new `(can-update? instance changes)` method to the permissions interface a while back... this sort of thing could be a good place to use it. We have this sort of "if this column is changing, check these permissions" code all over the place in the API endpoints. `can-update?` is supposed to be a consistent way of checking that, similar to how we pretty consistently use `can-read?` and `can-write?`...

If I had infinite refactor time I'd move all the stuff into `can-update?` and use it everywhere. But since I don't I've been trying to migrate stuff over piecemeal and use it for new stuff as it comes up.

Up to you. I'm not too bothered either way, just thought I'd mention it in case you think it makes sense